Podman and Gunicorn

KaratekHD opened the pull request:
This adds support for deploying the API using Podman.

why-not-try-calmer commented on deploying.md:
```diff
@@ -0,0 +1,37 @@
+### Using prebuilt containers
+1. Create a .env file as described in [configuration.md](configuration.md).
+2. Make sure you have Podman installed: `sudo zypper install podman`
+3. Run the container: `podman run --env-file /path/to/env/file -p 80:8000 registry.opensuse.org/home/karatekhd/defrag/containers/defrag:latest`. Replace the `/path/to/env/file` with the actual path tp your env file.
```
Forwarding 80 to 8000 won't work on a rootless podman, so we might want to use another port instead of 80.

KaratekHD replied:
That's true, I'll change it.

KaratekHD replied:
Fixed in 2807307

why-not-try-calmer commented on deploying.md:
```diff
@@ -0,0 +1,37 @@
+### Using prebuilt containers
+1. Create a .env file as described in [configuration.md](configuration.md).
+2. Make sure you have Podman installed: `sudo zypper install podman`
+3. Run the container: `podman run --env-file /path/to/env/file -p 80:8000 registry.opensuse.org/home/karatekhd/defrag/containers/defrag:latest`. Replace the `/path/to/env/file` with the actual path tp your env file.
```
Also this will trigger change to `defrag/__init__.py`, won't it? Do we still need the `config` parser after this change?

KaratekHD replied:
No, it won't. This parameter does the same creating a `.env` file does. It just passes it to the container so that it can read the variables specified in the env file.

KaratekHD replied:
We still need the config parser though.

why-not-try-calmer replied:
Alright then.

why-not-try-calmer commented on Dockerfile:
```diff
@@ -0,0 +1,29 @@
+
+FROM opensuse/tumbleweed
+
+ENV ENV "True"
```
It's the first time I see this. Why is it required?

KaratekHD replied:
This just sets the environment variable `ENV` to true. We could also remove it, but this would require `ENV=True` to be in the .env file. It's just there for convenience 😉

why-not-try-calmer replied:
Okay.

why-not-try-calmer reviewed (changes requested):
More like "Request info" instead of "Request changes" :)

why-not-try-calmer commented on setup.py:
```diff
@@ -0,0 +1,20 @@
+import os
+from setuptools import setup
+
+setup(
```
As a matter of fact I don't think there is a use case for using `setuptools`. This tool is typically used to make Python modules consumable as a library or an installable program, and as far as I know we are not quite in the business of providing this app as an installable, let alone a standalone library. 

In other words, I think this tool is going to far beyond the goal of "configuring the app in view of deploying it and testing it in production", which is the goal we are currently pursuing, if I am not mistaken.

KaratekHD replied:
I disagree. For building new containers automatically, there is a quite simple workflow on OBS:
1. get the current sources and build an RPM
2. once the RPM is built, the container is rebuilt using the new package.
By using setuptools, this is a really simple procedure.
All we need is this part in our .spec file:
```specfile
%build
%python_build

%install
%python_install
%python_expand %fdupes %{buildroot}%{$python_sitelib}
```

KaratekHD replied:
Note to myself: Comment this

why-not-try-calmer commented on Dockerfile:
```diff
@@ -0,0 +1,29 @@
+ENV ENV "True"
+EXPOSE 8000
+
+RUN zypper ar https://download.opensuse.org/repositories/home:/KaratekHD:/defrag/openSUSE_Tumbleweed/ defrag
```
Is it correct to say that if any of the following packages:
- atoma
- fastapi
- lunr
- mmh3
- pottery
- typing

gets an update, we will need to rebuild the corresponding package from OBS to apply the update? If yes, I think it's not very sustainable for two developers, is it? While if we simply installed from the requirements, this problem would not occur. What problem would occur if we installed from requirements that justifies re-packaging all these packages ourselves?

KaratekHD replied:
We'll, it's a matter of running `py2pack fetch atoma` and `py2pack generate atoma -f python-atoma.spec` - I can easily do this alone, I did these libs in like half an hour. 
The issue with installing from requirements.txt is that we can not automate it in OBS, meaning that we would need to rebuild the containers by hand. Since currently we have hardcoded versions in the requirements.txt anyway, I think it is fine to update the packages on OBS. I can do it so you don't have to worry ;)

KaratekHD replied:
If you want to further discuss this, I'd suggest having a small meeting about it.

why-not-try-calmer replied:
The problem is that you will need to remember that any of these packages *might* have a pending update and to run the commands. This means that you are now shifting an essential part of the workflow away from the codebase to this home repo. It's WAY too early for me. I want (me / the devs)  to be able to develop with their eyes focused on the codebase, automatically get the latest update without the overhead of having to remember to update some remote repo, to which they may not even have access. Again, we are conflating two concerns here:
- going to test-production (the current goal as per the recent meetings) with a containerized version of the application
- packaging (the long-term goal which you have already made a lot of progress with) in view of distributing the app as a module / library

I want to to be able to focus on the first objective, the rest can wait another PR. There is no need to lump together these two separate goals, especially when we have 0 integration test and a lot more testing in production to do...

KaratekHD replied:
Sure, we can skip packaging for now. But this would also mean not to use Podman for the deployment, as we are not able to build the containers on OBS otherwise. We can of course do that, write a simple .service file, install all the requirements from the file and run it with systemd. It's not fancy, but it is a solution.

KaratekHD replied:
If you're concern is that we could forget updating these packages, I can write `_service` files for the package so that the process gets automated

why-not-try-calmer replied:
> Sure, we can skip packaging for now. But this would also mean not to use Podman for the deployment, as we are not able to build the containers on OBS otherwise. We can of course do that, write a simple .service file, install all the requirements from the file and run it with systemd. It's not fancy, but it is a solution.

Well, we don't _have_ to use OBS for now. We can simply build from podman (`RUN pip install ...` and/or `RUN pipenv install ...`) and deploy from there. That's super simple and you get updates automatically whenever you rebuild.

I mean, all the good work you've done is not lost, it's just too early if the goal is to test the running code. ^^

KaratekHD replied:
Well, the updates for the deployment wouldn't need to occur everytime we push to dev. We probably should merge dev into main soon, build the package from main and deploy that. That way we can keep what we already have and develop without worries in dev. I mean, I would need to do the updates on the server side anyway, so I can as well trigger a service run for OBS then.

I honestly don't see how it is different to rebuild the package from rebuilding the container, apart from the point that using the packages is the more future proofed and cleaner way.

KaratekHD committed:
Add Podman support

Update requirements

Add Dockerfile

add setup.py

fix setup.py

support for Podman

change default port for podman

KaratekHD committed:
Don't use RPM packages

Clone sources in Dockerfile instead of installing from OBS

clean Dockerfile

Disable setup.py

KaratekHD committed:
Revert "Don't use RPM packages"

This reverts commit 31bd23361e2269713a372273e1e8f768a2b6efdf.

KaratekHD commented:
@why-not-try-calmer Please check this out ;)

why-not-try-calmer commented on Dockerfile:
```diff
@@ -0,0 +1,29 @@
+RUN zypper --gpg-auto-import-keys ref
+RUN zypper --non-interactive install python38-defrag-api
+
+CMD ["python3.8", "-m", "defrag"]
```
Could we try from the get go with gunicorn ([example](https://github.com/openSUSE/defrag-api/blob/c97b50580249955c611a5b08b4d0caf9e217fd49/Dockerfile#L11)) so that we don't have bad surprises later? Notice that this implies that `__main.py__` be renamed to `main.py` (a good thing in my opinion).

why-not-try-calmer reviewed (commented):
I wish to see the gunicorn added, but it's not a "blocker" for me :)

thunderbirdtr commented:
I wish to see gunicorn and I see as a blocker because, it is a feature already exists and not gonna remove anyway so we don't wanna lose the ability already exist so let's fix that first then merge it.

KaratekHD committed:
merge latest master in karatek/work/container

KaratekHD committed:
add opengm to setup.py

KaratekHD committed:
use multiple config files

KaratekHD committed:
add support for gunicorn

KaratekHD committed:
update deploying docs

KaratekHD commented:
I'm happy with this. Everything gets build on OBS and the container images use gunicorn. After this gets merged, please remember to add new python packages to setup.py.

why-not-try-calmer commented:
It looks like the Dockerfile from this repo does not match the Dockerfile at https://build.opensuse.org/package/view_file/openSUSE:infrastructure:defrag:containers/defrag-container/Dockerfile?expand=1. Would it be possible to have them match?

why-not-try-calmer commented:
> Sure, I forgot that. I'll replace the Dockerfile in this repo with the one from build-o.o. Apart from that, everything good?

Well, probably the docs need to be updated accordingly. I'd like to test after you've replaced the Dockerfile, and then I'll confirm!

KaratekHD committed:
update Dockerfile

KaratekHD commented:
Works fine for me when running as root:
<details>
<summary>Output</summary>

```shell
$ sudo podman build . -t defrag
STEP 1/7: FROM opensuse/tumbleweed
STEP 2/7: ENV ENV "True"
--> Using cache 4d32eaf193f81043cd8fec84e42bfedf6c74eb7e1af582adff93a2411b92bd02
--> 4d32eaf193f
STEP 3/7: EXPOSE 8000
--> Using cache 719c78418f59907c7da624a6a798bf6b5287111c3b7272d4fb5c2139e61095fb
--> 719c78418f5
STEP 4/7: RUN zypper ar https://download.opensuse.org/repositories/openSUSE:infrastructure:defrag/openSUSE_Tumbleweed/ defrag
Adding repository 'defrag' [......done]
Repository 'defrag' successfully added

URI         : https://download.opensuse.org/repositories/openSUSE:infrastructure:defrag/openSUSE_Tumbleweed/
Enabled     : Yes
GPG Check   : Yes
Autorefresh : No
Priority    : 99 (default priority)

Repository priorities are without effect. All enabled repositories share the same priority.
--> e671b92a3c9
STEP 5/7: RUN zypper --gpg-auto-import-keys ref
Retrieving repository 'defrag' metadata [..

Automatically importing the following key:

  Repository:       defrag
  Key Fingerprint:  7E8B 87D3 15BF 9661 8A3B 297B F9A8 B79B EAE4 FD92
  Key Name:         openSUSE OBS Project <openSUSE@build.opensuse.org>
  Key Algorithm:    DSA 1024
  Key Created:      Mon Jul 26 14:52:26 2021
  Key Expires:      Wed Oct  4 14:52:26 2023
  Rpm Name:         gpg-pubkey-eae4fd92-60fecc2a



    Note: A GPG pubkey is clearly identified by it's fingerprint. Do not rely the keys name. If you
    are not sure whether the presented key is authentic, ask the repository provider or check his
    web site. Many provider maintain a web page showing the fingerprints of the GPG keys they are
    using.
done]
Building repository 'defrag' cache [....done]
Retrieving repository 'openSUSE-Tumbleweed-Non-Oss' metadata [..done]
Building repository 'openSUSE-Tumbleweed-Non-Oss' cache [....done]
Retrieving repository 'openSUSE-Tumbleweed-Oss' metadata [................done]
Building repository 'openSUSE-Tumbleweed-Oss' cache [....done]
Retrieving repository 'openSUSE-Tumbleweed-Update' metadata [.done]
Building repository 'openSUSE-Tumbleweed-Update' cache [....done]
All repositories have been refreshed.
--> da0141143b8
STEP 6/7: RUN zypper --non-interactive install --recommends python38-defrag-api python38-gunicorn
Loading repository data...
Reading installed packages...
Resolving package dependencies...

The following 2 packages are going to be upgraded:
  glibc glibc-locale-base

The following 14 recommended packages were automatically selected:
  login_defs logrotate python38 python38-PySocks python38-PyYAML python38-aiodns python38-brotlipy python38-cryptography python38-curses python38-dbm python38-hiredis python38-idna python38-pip python38-pyOpenSSL

The following 5 packages are suggested, but will not be installed:
  python38-gevent python38-setproctitle python38-tornado python38-tornado5 python38-tornado6

The following 102 NEW packages are going to be installed:
  chkstat glibc-extra libaudit1 libbrotlienc1 libcrypt1 libeconf0 libexpat1 libgdbm6 libgdbm_compat4 libhiredis1_0_0 libhttp_parser2_9 libmpdec3 libnsl3 libpython3_8-1_0 libsemanage-conf libsemanage2 libsepol2 libtirpc-netconfig libtirpc3 libuuid1 libuv1 libxslt1 login_defs logrotate pam pam_unix permissions permissions-config python38 python38-PyJWT python38-PySocks python38-PyYAML python38-aiodns python38-aiohttp python38-anyio python38-apipkg python38-appdirs python38-asgiref python38-asn1crypto python38-async_timeout python38-atoma python38-attrs python38-base python38-beautifulsoup4 python38-blinker python38-brotlipy python38-bugzilla python38-certifi python38-cffi python38-chardet python38-click python38-cryptography python38-cssselect python38-curses python38-dbm python38-defrag-api python38-defusedxml python38-fastapi python38-future python38-gunicorn python38-h11 python38-hiredis python38-httptools python38-idna python38-iniconfig python38-lunr python38-lxml python38-mmh3 python38-multidict python38-oauthlib python38-ordered-set python38-packaging python38-pip python38-pottery python38-py python38-pyOpenSSL python38-pyasn1 python38-pycares python38-pycparser python38-pydantic python38-pyparsing python38-python-dateutil python38-python-dotenv python38-python-twitter python38-redis python38-requests python38-requests-oauthlib python38-setuptools python38-six python38-sniffio python38-soupsieve python38-starlette python38-typing_extensions python38-urllib3 python38-uvicorn python38-uvloop python38-websockets python38-yarl redis shadow shared-python-startup sysuser-shadow

2 packages to upgrade, 102 new.
Overall download size: 31.7 MiB. Already cached: 0 B. After the operation, additional 107.4 MiB will be used.
Continue? [y/n/v/...? shows all options] (y): y
Retrieving package glibc-2.34-1.2.x86_64 (1/104),   1.8 MiB (  5.8 MiB unpacked)
Retrieving: glibc-2.34-1.2.x86_64.rpm [..done (1.8 MiB/s)]
Retrieving package libsemanage-conf-3.2-2.2.x86_64 (2/104),  19.8 KiB (  2.3 KiB unpacked)
Retrieving: libsemanage-conf-3.2-2.2.x86_64.rpm [done]
Retrieving package libtirpc-netconfig-1.3.2-2.1.x86_64 (3/104),  29.3 KiB (  1.9 KiB unpacked)
Retrieving: libtirpc-netconfig-1.3.2-2.1.x86_64.rpm [done]
Retrieving package login_defs-4.9-1.1.noarch (4/104),  37.7 KiB ( 18.6 KiB unpacked)
Retrieving: login_defs-4.9-1.1.noarch.rpm [done]
Retrieving package shared-python-startup-0.1-6.5.noarch (5/104),  12.9 KiB ( 13.5 KiB unpacked)
Retrieving: shared-python-startup-0.1-6.5.noarch.rpm [done]
Retrieving package glibc-locale-base-2.34-1.2.x86_64 (6/104),   2.2 MiB ( 12.3 MiB unpacked)
Retrieving: glibc-locale-base-2.34-1.2.x86_64.rpm [.done (1.6 MiB/s)]
Retrieving package glibc-extra-2.34-1.2.x86_64 (7/104), 216.6 KiB ( 27.6 KiB unpacked)
Retrieving: glibc-extra-2.34-1.2.x86_64.rpm [.done (1.0 MiB/s)]
Retrieving package logrotate-3.18.1-1.2.x86_64 (8/104),  78.9 KiB (159.0 KiB unpacked)
Retrieving: logrotate-3.18.1-1.2.x86_64.rpm [done]
Retrieving package libxslt1-1.1.34-3.2.x86_64 (9/104), 177.1 KiB (356.6 KiB unpacked)
Retrieving: libxslt1-1.1.34-3.2.x86_64.rpm [done]
Retrieving package libuv1-1.42.0-1.2.x86_64 (10/104), 127.5 KiB (199.4 KiB unpacked)
Retrieving: libuv1-1.42.0-1.2.x86_64.rpm [done]
Retrieving package libuuid1-2.36.2-3.2.x86_64 (11/104), 150.3 KiB ( 30.2 KiB unpacked)
Retrieving: libuuid1-2.36.2-3.2.x86_64.rpm [.done]
Retrieving package libtirpc3-1.3.2-2.1.x86_64 (12/104), 110.4 KiB (217.9 KiB unpacked)
Retrieving: libtirpc3-1.3.2-2.1.x86_64.rpm [done]
Retrieving package libsepol2-3.2-3.2.x86_64 (13/104), 283.1 KiB (713.4 KiB unpacked)
Retrieving: libsepol2-3.2-3.2.x86_64.rpm [done]
Retrieving package libmpdec3-2.5.1-2.2.x86_64 (14/104),  80.9 KiB (193.0 KiB unpacked)
Retrieving: libmpdec3-2.5.1-2.2.x86_64.rpm [.done (1.3 MiB/s)]
Retrieving package libhttp_parser2_9-2.9.4-1.7.x86_64 (15/104),  26.5 KiB ( 52.4 KiB unpacked)
Retrieving: libhttp_parser2_9-2.9.4-1.7.x86_64.rpm [done]
Retrieving package libhiredis1_0_0-1.0.0-2.6.x86_64 (16/104),  44.6 KiB ( 97.4 KiB unpacked)
Retrieving: libhiredis1_0_0-1.0.0-2.6.x86_64.rpm [done]
Retrieving package libgdbm6-1.20-1.2.x86_64 (17/104),  57.4 KiB (105.5 KiB unpacked)
Retrieving: libgdbm6-1.20-1.2.x86_64.rpm [done]
Retrieving package libexpat1-2.4.1-1.2.x86_64 (18/104), 110.6 KiB (206.3 KiB unpacked)
Retrieving: libexpat1-2.4.1-1.2.x86_64.rpm [done]
Retrieving package libeconf0-0.4.1+git20210709.cf671f2-1.2.x86_64 (19/104),  33.0 KiB ( 39.7 KiB unpacked)
Retrieving: libeconf0-0.4.1+git20210709.cf671f2-1.2.x86_64.rpm [done]
Retrieving package libcrypt1-4.4.25-1.2.x86_64 (20/104), 114.5 KiB (264.4 KiB unpacked)
Retrieving: libcrypt1-4.4.25-1.2.x86_64.rpm [.done (1.5 MiB/s)]
Retrieving package libbrotlienc1-1.0.9-1.5.x86_64 (21/104), 239.8 KiB (578.0 KiB unpacked)
Retrieving: libbrotlienc1-1.0.9-1.5.x86_64.rpm [done]
Retrieving package libaudit1-3.0.5-1.1.x86_64 (22/104),  98.5 KiB (114.8 KiB unpacked)
Retrieving: libaudit1-3.0.5-1.1.x86_64.rpm [done]
Retrieving package chkstat-1550_20210901-29.2.x86_64 (23/104), 412.5 KiB (  1.2 MiB unpacked)
Retrieving: chkstat-1550_20210901-29.2.x86_64.rpm [.done (1.7 MiB/s)]
Retrieving package libnsl3-2.0.0-1.2.x86_64 (24/104),  30.0 KiB ( 62.0 KiB unpacked)
Retrieving: libnsl3-2.0.0-1.2.x86_64.rpm [done]
Retrieving package libgdbm_compat4-1.20-1.2.x86_64 (25/104),  33.1 KiB ( 52.5 KiB unpacked)
Retrieving: libgdbm_compat4-1.20-1.2.x86_64.rpm [done]
Retrieving package libsemanage2-3.2-2.2.x86_64 (26/104), 108.3 KiB (267.4 KiB unpacked)
Retrieving: libsemanage2-3.2-2.2.x86_64.rpm [done]
Retrieving package permissions-config-1550_20210901-29.2.x86_64 (27/104),  60.1 KiB ( 46.9 KiB unpacked)
Retrieving: permissions-config-1550_20210901-29.2.x86_64.rpm [done]
Retrieving package python38-base-3.8.12-1.3.x86_64 (28/104),   8.7 MiB ( 33.0 MiB unpacked)
Retrieving: python38-base-3.8.12-1.3.x86_64.rpm [.....done (2.0 MiB/s)]
Retrieving package permissions-20210901.1550-29.2.x86_64 (29/104),  24.3 KiB (    0   B unpacked)
Retrieving: permissions-20210901.1550-29.2.x86_64.rpm [done]
Retrieving package libpython3_8-1_0-3.8.12-1.3.x86_64 (30/104),   1.2 MiB (  3.2 MiB unpacked)
Retrieving: libpython3_8-1_0-3.8.12-1.3.x86_64.rpm [.done (1.6 MiB/s)]
Retrieving package pam_unix-1.5.2-1.1.x86_64 (31/104),  71.2 KiB ( 54.8 KiB unpacked)
Retrieving: pam_unix-1.5.2-1.1.x86_64.rpm [done]
Retrieving package python38-six-1.16.0-2.2.noarch (32/104),  41.7 KiB (100.8 KiB unpacked)
Retrieving: python38-six-1.16.0-2.2.noarch.rpm [done]
Retrieving package python38-pyparsing-2.4.7-3.2.noarch (33/104), 201.6 KiB (872.7 KiB unpacked)
Retrieving: python38-pyparsing-2.4.7-3.2.noarch.rpm [.done (1.6 MiB/s)]
Retrieving package python38-ordered-set-4.0.2-1.2.noarch (34/104),  21.4 KiB ( 44.7 KiB unpacked)
Retrieving: python38-ordered-set-4.0.2-1.2.noarch.rpm [done]
Retrieving package python38-appdirs-1.4.4-3.2.noarch (35/104),  24.1 KiB ( 84.0 KiB unpacked)
Retrieving: python38-appdirs-1.4.4-3.2.noarch.rpm [done]
Retrieving package pam-1.5.2-1.1.x86_64 (36/104), 440.7 KiB (  1.6 MiB unpacked)
Retrieving: pam-1.5.2-1.1.x86_64.rpm [done]
Retrieving package python38-packaging-20.9-2.2.noarch (37/104),  77.3 KiB (284.9 KiB unpacked)
Retrieving: python38-packaging-20.9-2.2.noarch.rpm [.done]
Retrieving package shadow-4.9-1.1.x86_64 (38/104), 686.6 KiB (  3.2 MiB unpacked)
Retrieving: shadow-4.9-1.1.x86_64.rpm [done]
Retrieving package python38-setuptools-57.4.0-1.2.noarch (39/104), 1004.7 KiB (  5.2 MiB unpacked)
Retrieving: python38-setuptools-57.4.0-1.2.noarch.rpm [.done (1.7 MiB/s)]
Retrieving package sysuser-shadow-3.1-3.2.noarch (40/104),  13.1 KiB (  2.5 KiB unpacked)
Retrieving: sysuser-shadow-3.1-3.2.noarch.rpm [done]
Retrieving package python38-pip-20.2.4-1.8.noarch (41/104),   2.0 MiB ( 10.4 MiB unpacked)
Retrieving: python38-pip-20.2.4-1.8.noarch.rpm [.done (1.6 MiB/s)]
Retrieving package redis-6.2.5-1.2.x86_64 (42/104),   1.3 MiB (  4.7 MiB unpacked)
Retrieving: redis-6.2.5-1.2.x86_64.rpm [.done (241.4 KiB/s)]
Retrieving package python38-3.8.12-1.2.x86_64 (43/104), 104.1 KiB (160.8 KiB unpacked)
Retrieving: python38-3.8.12-1.2.x86_64.rpm [done]
Retrieving package python38-dbm-3.8.12-1.2.x86_64 (44/104),  95.1 KiB ( 85.0 KiB unpacked)
Retrieving: python38-dbm-3.8.12-1.2.x86_64.rpm [.done]
Retrieving package python38-curses-3.8.12-1.2.x86_64 (45/104), 127.0 KiB (186.3 KiB unpacked)
Retrieving: python38-curses-3.8.12-1.2.x86_64.rpm [done]
Retrieving package python38-websockets-9.1-2.1.x86_64 (46/104), 132.0 KiB (698.5 KiB unpacked)
Retrieving: python38-websockets-9.1-2.1.x86_64.rpm [done]
Retrieving package python38-uvloop-0.16.0-2.2.x86_64 (47/104), 560.7 KiB (  1.8 MiB unpacked)
Retrieving: python38-uvloop-0.16.0-2.2.x86_64.rpm [.done (1.6 MiB/s)]
Retrieving package python38-starlette-0.16.0-2.1.noarch (48/104), 128.5 KiB (568.7 KiB unpacked)
Retrieving: python38-starlette-0.16.0-2.1.noarch.rpm [done]
Retrieving package python38-soupsieve-2.2.1-1.2.noarch (49/104),  73.8 KiB (295.1 KiB unpacked)
Retrieving: python38-soupsieve-2.2.1-1.2.noarch.rpm [done]
Retrieving package python38-sniffio-1.2.0-4.4.noarch (50/104),  21.4 KiB ( 31.1 KiB unpacked)
Retrieving: python38-sniffio-1.2.0-4.4.noarch.rpm [.done]
Retrieving package python38-python-dateutil-2.8.2-1.2.noarch (51/104), 344.0 KiB (888.4 KiB unpacked)
Retrieving: python38-python-dateutil-2.8.2-1.2.noarch.rpm [done]
Retrieving package python38-pycparser-2.20-1.11.noarch (52/104), 226.5 KiB (  1.3 MiB unpacked)
Retrieving: python38-pycparser-2.20-1.11.noarch.rpm [done]
Retrieving package python38-pyasn1-0.4.8-2.2.noarch (53/104), 172.8 KiB (919.3 KiB unpacked)
Retrieving: python38-pyasn1-0.4.8-2.2.noarch.rpm [.done]
Retrieving package python38-multidict-5.1.0-1.7.x86_64 (54/104),  51.3 KiB (150.7 KiB unpacked)
Retrieving: python38-multidict-5.1.0-1.7.x86_64.rpm [done]
Retrieving package python38-iniconfig-1.1.1-1.5.noarch (55/104),  16.1 KiB ( 21.8 KiB unpacked)
Retrieving: python38-iniconfig-1.1.1-1.5.noarch.rpm [done]
Retrieving package python38-idna-3.2-1.2.noarch (56/104),  99.7 KiB (699.9 KiB unpacked)
Retrieving: python38-idna-3.2-1.2.noarch.rpm [done]
Retrieving package python38-httptools-0.1.1-1.7.x86_64 (57/104),  43.5 KiB (102.9 KiB unpacked)
Retrieving: python38-httptools-0.1.1-1.7.x86_64.rpm [done]
Retrieving package python38-hiredis-1.1.0-1.5.x86_64 (58/104),  20.7 KiB ( 32.2 KiB unpacked)
Retrieving: python38-hiredis-1.1.0-1.5.x86_64.rpm [done]
Retrieving package python38-h11-0.12.0-1.4.noarch (59/104),  95.3 KiB (390.1 KiB unpacked)
Retrieving: python38-h11-0.12.0-1.4.noarch.rpm [.done]
Retrieving package python38-gunicorn-20.1.0-1.2.noarch (60/104), 169.5 KiB (664.0 KiB unpacked)
Retrieving: python38-gunicorn-20.1.0-1.2.noarch.rpm [done]
Retrieving package python38-future-0.18.2-3.4.noarch (61/104), 757.6 KiB (  3.7 MiB unpacked)
Retrieving: python38-future-0.18.2-3.4.noarch.rpm [.done (1.9 MiB/s)]
Retrieving package python38-defusedxml-0.7.1-2.2.noarch (62/104),  43.5 KiB (154.7 KiB unpacked)
Retrieving: python38-defusedxml-0.7.1-2.2.noarch.rpm [done]
Retrieving package python38-cssselect-1.1.0-1.8.noarch (63/104),  42.3 KiB (153.7 KiB unpacked)
Retrieving: python38-cssselect-1.1.0-1.8.noarch.rpm [done]
Retrieving package python38-click-8.0.1-2.2.noarch (64/104), 196.4 KiB (951.1 KiB unpacked)
Retrieving: python38-click-8.0.1-2.2.noarch.rpm [done]
Retrieving package python38-chardet-4.0.0-1.5.noarch (65/104), 206.3 KiB (  1.6 MiB unpacked)
Retrieving: python38-chardet-4.0.0-1.5.noarch.rpm [.done (5.4 KiB/s)]
Retrieving package python38-certifi-2021.5.30-1.2.noarch (66/104),  18.6 KiB (  8.3 KiB unpacked)
Retrieving: python38-certifi-2021.5.30-1.2.noarch.rpm [done]
Retrieving package python38-blinker-1.4-5.6.noarch (67/104),  30.7 KiB ( 92.9 KiB unpacked)
Retrieving: python38-blinker-1.4-5.6.noarch.rpm [done]
Retrieving package python38-attrs-21.2.0-1.2.noarch (68/104), 120.6 KiB (464.8 KiB unpacked)
Retrieving: python38-attrs-21.2.0-1.2.noarch.rpm [done]
Retrieving package python38-async_timeout-3.0.1-2.5.noarch (69/104),  19.0 KiB ( 25.6 KiB unpacked)
Retrieving: python38-async_timeout-3.0.1-2.5.noarch.rpm [done]
Retrieving package python38-asn1crypto-1.4.0-2.2.noarch (70/104), 192.0 KiB (  1.2 MiB unpacked)
Retrieving: python38-asn1crypto-1.4.0-2.2.noarch.rpm [done]
Retrieving package python38-asgiref-3.4.1-1.2.noarch (71/104),  50.9 KiB (159.1 KiB unpacked)
Retrieving: python38-asgiref-3.4.1-1.2.noarch.rpm [.done (5.4 KiB/s)]
Retrieving package python38-apipkg-1.5-4.5.noarch (72/104),  20.2 KiB ( 29.3 KiB unpacked)
Retrieving: python38-apipkg-1.5-4.5.noarch.rpm [done]
Retrieving package python38-PyYAML-5.4.1-1.6.x86_64 (73/104), 112.9 KiB (553.7 KiB unpacked)
Retrieving: python38-PyYAML-5.4.1-1.6.x86_64.rpm [done]
Retrieving package python38-PySocks-1.7.1-1.9.noarch (74/104),  33.5 KiB ( 97.2 KiB unpacked)
Retrieving: python38-PySocks-1.7.1-1.9.noarch.rpm [done]
Retrieving package python38-cffi-1.14.6-1.2.x86_64 (75/104), 330.6 KiB (  1.2 MiB unpacked)
Retrieving: python38-cffi-1.14.6-1.2.x86_64.rpm [done]
Retrieving package python38-yarl-1.6.3-3.4.x86_64 (76/104),  84.8 KiB (259.3 KiB unpacked)
Retrieving: python38-yarl-1.6.3-3.4.x86_64.rpm [.done (760.7 KiB/s)]
Retrieving package python38-anyio-3.3.1-1.2.noarch (77/104), 163.5 KiB (784.8 KiB unpacked)
Retrieving: python38-anyio-3.3.1-1.2.noarch.rpm [done]
Retrieving package python38-redis-3.5.3-3.2.noarch (78/104), 174.3 KiB (749.3 KiB unpacked)
Retrieving: python38-redis-3.5.3-3.2.noarch.rpm [done]
Retrieving package python38-lxml-4.6.3-3.2.x86_64 (79/104),   1.3 MiB (  4.3 MiB unpacked)
Retrieving: python38-lxml-4.6.3-3.2.x86_64.rpm [.done (2.0 MiB/s)]
Retrieving package python38-python-dotenv-0.15.0-1.6.noarch (80/104),  38.7 KiB ( 95.4 KiB unpacked)
Retrieving: python38-python-dotenv-0.15.0-1.6.noarch.rpm [done]
Retrieving package python38-py-1.10.0-1.5.noarch (81/104), 182.9 KiB (724.3 KiB unpacked)
Retrieving: python38-py-1.10.0-1.5.noarch.rpm [.done (1.7 MiB/s)]
Retrieving package python38-pycares-4.0.0-1.2.x86_64 (82/104),  92.4 KiB (280.7 KiB unpacked)
Retrieving: python38-pycares-4.0.0-1.2.x86_64.rpm [done]
Retrieving package python38-cryptography-3.3.2-2.4.x86_64 (83/104), 472.0 KiB (  2.6 MiB unpacked)
Retrieving: python38-cryptography-3.3.2-2.4.x86_64.rpm [done]
Retrieving package python38-brotlipy-0.7.0-5.9.x86_64 (84/104),  27.2 KiB ( 75.5 KiB unpacked)
Retrieving: python38-brotlipy-0.7.0-5.9.x86_64.rpm [.done]
Retrieving package python38-beautifulsoup4-4.9.3-1.9.noarch (85/104), 219.9 KiB (  1.1 MiB unpacked)
Retrieving: python38-beautifulsoup4-4.9.3-1.9.noarch.rpm [done]
Retrieving package python38-aiodns-3.0.0-1.2.noarch (86/104),  18.6 KiB ( 23.6 KiB unpacked)
Retrieving: python38-aiodns-3.0.0-1.2.noarch.rpm [done]
Retrieving package python38-pyOpenSSL-20.0.1-1.4.noarch (87/104), 115.6 KiB (579.0 KiB unpacked)
Retrieving: python38-pyOpenSSL-20.0.1-1.4.noarch.rpm [done]
Retrieving package python38-PyJWT-2.1.0-1.2.noarch (88/104),  54.2 KiB (154.2 KiB unpacked)
Retrieving: python38-PyJWT-2.1.0-1.2.noarch.rpm [.done (998.0 KiB/s)]
Retrieving package python38-urllib3-1.26.6-1.2.noarch (89/104), 217.2 KiB (708.7 KiB unpacked)
Retrieving: python38-urllib3-1.26.6-1.2.noarch.rpm [done]
Retrieving package python38-oauthlib-3.1.1-1.2.noarch (90/104), 202.1 KiB (  1.3 MiB unpacked)
Retrieving: python38-oauthlib-3.1.1-1.2.noarch.rpm [done]
Retrieving package python38-requests-2.25.1-4.2.noarch (91/104), 167.9 KiB (503.3 KiB unpacked)
Retrieving: python38-requests-2.25.1-4.2.noarch.rpm [.done (1.3 MiB/s)]
Retrieving package python38-requests-oauthlib-1.3.0-2.1.noarch (92/104),  49.1 KiB (157.6 KiB unpacked)
Retrieving: python38-requests-oauthlib-1.3.0-2.1.noarch.rpm [done]
Retrieving package python38-bugzilla-3.0.2-1.6.noarch (93/104), 113.4 KiB (492.9 KiB unpacked)
Retrieving: python38-bugzilla-3.0.2-1.6.noarch.rpm [done]
Retrieving package python38-python-twitter-3.5-1.12.noarch (94/104), 109.9 KiB (711.6 KiB unpacked)
Retrieving: python38-python-twitter-3.5-1.12.noarch.rpm [done]
Retrieving package python38-typing_extensions-3.10.0.2-1.1.noarch (95/104),  49.4 KiB (271.9 KiB unpacked)
Retrieving: python38-typing_extensions-3.10.0.2-1.1.noarch.rpm [.done]
Retrieving package python38-mmh3-3.0.0-1.1.x86_64 (96/104),  17.6 KiB ( 34.8 KiB unpacked)
Retrieving: python38-mmh3-3.0.0-1.1.x86_64.rpm [done]
Retrieving package python38-lunr-0.6.0-1.1.noarch (97/104),  66.5 KiB (246.8 KiB unpacked)
Retrieving: python38-lunr-0.6.0-1.1.noarch.rpm [done]
Retrieving package python38-fastapi-0.68.1-1.1.noarch (98/104), 101.0 KiB (439.2 KiB unpacked)
Retrieving: python38-fastapi-0.68.1-1.1.noarch.rpm [.done (1.3 MiB/s)]
Retrieving package python38-atoma-0.0.17-1.1.noarch (99/104),  32.4 KiB (106.8 KiB unpacked)
Retrieving: python38-atoma-0.0.17-1.1.noarch.rpm [done]
Retrieving package python38-pottery-1.3.5-1.1.noarch (100/104),  80.1 KiB (421.4 KiB unpacked)
Retrieving: python38-pottery-1.3.5-1.1.noarch.rpm [done]
Retrieving package python38-uvicorn-0.12.3-2.2.noarch (101/104),  93.9 KiB (339.2 KiB unpacked)
Retrieving: python38-uvicorn-0.12.3-2.2.noarch.rpm [done]
Retrieving package python38-pydantic-1.8.2-1.2.noarch (102/104), 228.2 KiB (  1.1 MiB unpacked)
Retrieving: python38-pydantic-1.8.2-1.2.noarch.rpm [.done (1.6 MiB/s)]
Retrieving package python38-aiohttp-3.7.4-4.2.x86_64 (103/104), 480.8 KiB (  2.3 MiB unpacked)
Retrieving: python38-aiohttp-3.7.4-4.2.x86_64.rpm [done]
Retrieving package python38-defrag-api-0.1+git.20211005.5ab2621-6.1.x86_64 (104/104),  83.1 KiB (294.5 KiB unpacked)
Retrieving: python38-defrag-api-0.1+git.20211005.5ab2621-6.1.x86_64.rpm [.done]

Checking for file conflicts: [...........done]
(  1/104) Installing: glibc-2.34-1.2.x86_64 [...................done]
Additional rpm output:
/usr/sbin/ldconfig: /lib64/ld-linux-x86-64.so.2 is not a symbolic link



(  2/104) Installing: libsemanage-conf-3.2-2.2.x86_64 [.....done]
(  3/104) Installing: libtirpc-netconfig-1.3.2-2.1.x86_64 [....done]
(  4/104) Installing: login_defs-4.9-1.1.noarch [....done]
(  5/104) Installing: shared-python-startup-0.1-6.5.noarch [.....done]
(  6/104) Installing: glibc-locale-base-2.34-1.2.x86_64 [...................done]
(  7/104) Installing: glibc-extra-2.34-1.2.x86_64 [.....done]
(  8/104) Installing: logrotate-3.18.1-1.2.x86_64 [.........done]
(  9/104) Installing: libxslt1-1.1.34-3.2.x86_64 [.........done]
( 10/104) Installing: libuv1-1.42.0-1.2.x86_64 [.........done]
( 11/104) Installing: libuuid1-2.36.2-3.2.x86_64 [....done]
( 12/104) Installing: libtirpc3-1.3.2-2.1.x86_64 [..........done]
( 13/104) Installing: libsepol2-3.2-3.2.x86_64 [...........done]
( 14/104) Installing: libmpdec3-2.5.1-2.2.x86_64 [.........done]
( 15/104) Installing: libhttp_parser2_9-2.9.4-1.7.x86_64 [......done]
( 16/104) Installing: libhiredis1_0_0-1.0.0-2.6.x86_64 [.......done]
( 17/104) Installing: libgdbm6-1.20-1.2.x86_64 [......done]
( 18/104) Installing: libexpat1-2.4.1-1.2.x86_64 [.........done]
( 19/104) Installing: libeconf0-0.4.1+git20210709.cf671f2-1.2.x86_64 [.....done]
( 20/104) Installing: libcrypt1-4.4.25-1.2.x86_64 [...........done]
( 21/104) Installing: libbrotlienc1-1.0.9-1.5.x86_64 [............done]
( 22/104) Installing: libaudit1-3.0.5-1.1.x86_64 [.......done]
( 23/104) Installing: chkstat-1550_20210901-29.2.x86_64 [............done]
( 24/104) Installing: libnsl3-2.0.0-1.2.x86_64 [.....done]
( 25/104) Installing: libgdbm_compat4-1.20-1.2.x86_64 [.....done]
( 26/104) Installing: libsemanage2-3.2-2.2.x86_64 [...........done]
( 27/104) Installing: permissions-config-1550_20210901-29.2.x86_64 [.......done]
Additional rpm output:
Updating /etc/sysconfig/security ...
Checking permissions and ownerships - using the permissions files
        /usr/share/permissions/permissions
        /usr/share/permissions/permissions.easy
        /etc/permissions.local


( 28/104) Installing: python38-base-3.8.12-1.3.x86_64 [............done]
( 29/104) Installing: permissions-20210901.1550-29.2.x86_64 [...done]
( 30/104) Installing: libpython3_8-1_0-3.8.12-1.3.x86_64 [............done]
( 31/104) Installing: pam_unix-1.5.2-1.1.x86_64 [......done]
( 32/104) Installing: python38-six-1.16.0-2.2.noarch [.......done]
( 33/104) Installing: python38-pyparsing-2.4.7-3.2.noarch [............done]
( 34/104) Installing: python38-ordered-set-4.0.2-1.2.noarch [.......done]
( 35/104) Installing: python38-appdirs-1.4.4-3.2.noarch [........done]
( 36/104) Installing: pam-1.5.2-1.1.x86_64 [............done]
( 37/104) Installing: python38-packaging-20.9-2.2.noarch [...........done]
( 38/104) Installing: shadow-4.9-1.1.x86_64 [............done]
( 39/104) Installing: python38-setuptools-57.4.0-1.2.noarch [............done]
( 40/104) Installing: sysuser-shadow-3.1-3.2.noarch [....done]
( 41/104) Installing: python38-pip-20.2.4-1.8.noarch [............done]
( 42/104) Installing: redis-6.2.5-1.2.x86_64 [............done]
Additional rpm output:
/usr/sbin/useradd -r -c User for redis key-value store -d /var/lib/redis -U redis -s /usr/sbin/nologin
useradd: Warning: missing or non-executable shell '/usr/sbin/nologin'
See /usr/share/doc/packages/redis/README.SUSE to continue


( 43/104) Installing: python38-3.8.12-1.2.x86_64 [........done]
( 44/104) Installing: python38-dbm-3.8.12-1.2.x86_64 [.........done]
( 45/104) Installing: python38-curses-3.8.12-1.2.x86_64 [..........done]
( 46/104) Installing: python38-websockets-9.1-2.1.x86_64 [............done]
( 47/104) Installing: python38-uvloop-0.16.0-2.2.x86_64 [............done]
( 48/104) Installing: python38-starlette-0.16.0-2.1.noarch [............done]
( 49/104) Installing: python38-soupsieve-2.2.1-1.2.noarch [..........done]
( 50/104) Installing: python38-sniffio-1.2.0-4.4.noarch [.........done]
( 51/104) Installing: python38-python-dateutil-2.8.2-1.2.noarch [...........done]
( 52/104) Installing: python38-pycparser-2.20-1.11.noarch [............done]
( 53/104) Installing: python38-pyasn1-0.4.8-2.2.noarch [............done]
( 54/104) Installing: python38-multidict-5.1.0-1.7.x86_64 [..........done]
( 55/104) Installing: python38-iniconfig-1.1.1-1.5.noarch [.........done]
( 56/104) Installing: python38-idna-3.2-1.2.noarch [............done]
( 57/104) Installing: python38-httptools-0.1.1-1.7.x86_64 [.......done]
( 58/104) Installing: python38-hiredis-1.1.0-1.5.x86_64 [.....done]
( 59/104) Installing: python38-h11-0.12.0-1.4.noarch [...........done]
( 60/104) Installing: python38-gunicorn-20.1.0-1.2.noarch [............done]
( 61/104) Installing: python38-future-0.18.2-3.4.noarch [............done]
( 62/104) Installing: python38-defusedxml-0.7.1-2.2.noarch [..........done]
( 63/104) Installing: python38-cssselect-1.1.0-1.8.noarch [..........done]
( 64/104) Installing: python38-click-8.0.1-2.2.noarch [............done]
( 65/104) Installing: python38-chardet-4.0.0-1.5.noarch [............done]
( 66/104) Installing: python38-certifi-2021.5.30-1.2.noarch [.........done]
( 67/104) Installing: python38-blinker-1.4-5.6.noarch [..........done]
( 68/104) Installing: python38-attrs-21.2.0-1.2.noarch [...........done]
( 69/104) Installing: python38-async_timeout-3.0.1-2.5.noarch [.........done]
( 70/104) Installing: python38-asn1crypto-1.4.0-2.2.noarch [............done]
( 71/104) Installing: python38-asgiref-3.4.1-1.2.noarch [...........done]
( 72/104) Installing: python38-apipkg-1.5-4.5.noarch [.........done]
( 73/104) Installing: python38-PyYAML-5.4.1-1.6.x86_64 [...........done]
( 74/104) Installing: python38-PySocks-1.7.1-1.9.noarch [.......done]
( 75/104) Installing: python38-cffi-1.14.6-1.2.x86_64 [............done]
( 76/104) Installing: python38-yarl-1.6.3-3.4.x86_64 [..........done]
( 77/104) Installing: python38-anyio-3.3.1-1.2.noarch [............done]
( 78/104) Installing: python38-redis-3.5.3-3.2.noarch [............done]
( 79/104) Installing: python38-lxml-4.6.3-3.2.x86_64 [............done]
( 80/104) Installing: python38-python-dotenv-0.15.0-1.6.noarch [...........done]
( 81/104) Installing: python38-py-1.10.0-1.5.noarch [............done]
( 82/104) Installing: python38-pycares-4.0.0-1.2.x86_64 [...........done]
( 83/104) Installing: python38-cryptography-3.3.2-2.4.x86_64 [............done]
( 84/104) Installing: python38-brotlipy-0.7.0-5.9.x86_64 [.........done]
( 85/104) Installing: python38-beautifulsoup4-4.9.3-1.9.noarch [............done]
( 86/104) Installing: python38-aiodns-3.0.0-1.2.noarch [.........done]
( 87/104) Installing: python38-pyOpenSSL-20.0.1-1.4.noarch [...........done]
( 88/104) Installing: python38-PyJWT-2.1.0-1.2.noarch [..........done]
( 89/104) Installing: python38-urllib3-1.26.6-1.2.noarch [............done]
( 90/104) Installing: python38-oauthlib-3.1.1-1.2.noarch [............done]
( 91/104) Installing: python38-requests-2.25.1-4.2.noarch [............done]
( 92/104) Installing: python38-requests-oauthlib-1.3.0-2.1.noarch [...........done]
( 93/104) Installing: python38-bugzilla-3.0.2-1.6.noarch [...........done]
( 94/104) Installing: python38-python-twitter-3.5-1.12.noarch [...........done]
( 95/104) Installing: python38-typing_extensions-3.10.0.2-1.1.noarch [..........done]
( 96/104) Installing: python38-mmh3-3.0.0-1.1.x86_64 [.....done]
( 97/104) Installing: python38-lunr-0.6.0-1.1.noarch [............done]
( 98/104) Installing: python38-fastapi-0.68.1-1.1.noarch [............done]
( 99/104) Installing: python38-atoma-0.0.17-1.1.noarch [...........done]
(100/104) Installing: python38-pottery-1.3.5-1.1.noarch [...........done]
(101/104) Installing: python38-uvicorn-0.12.3-2.2.noarch [............done]
(102/104) Installing: python38-pydantic-1.8.2-1.2.noarch [............done]
(103/104) Installing: python38-aiohttp-3.7.4-4.2.x86_64 [............done]
(104/104) Installing: python38-defrag-api-0.1+git.20211005.5ab2621-6.1.x86_64 [............done]
Executing %posttrans script 'login_defs-4.9-1.1.noarch.rpm' [......done]
 
--> 4f60997d89f
STEP 7/7: CMD ["gunicorn", "-k", "uvicorn.workers.UvicornWorker", "defrag.__main__:app"]
COMMIT defrag
--> f997f24b9e2
Successfully tagged localhost/defrag:latest
f997f24b9e2cb62e377d5d42d15f06174a1eed5b32ed1d6421fd2f9b05d83585
```

</details>

As non-root user I can't build it either, but that's not a problem, as OBS builds as root. I'm gonna update the documentation though.

KaratekHD committed:
improve deploying docs
